File: lib/data/providers/official_api_provider.dart
import 'dart:convert';
import 'package:dio/dio.dart';
import '../../domain/entities/mosque.dart';
import '../../domain/entities/prayer_times.dart';
import '../../domain/entities/geo_location.dart';
import 'prayer_data_provider.dart';

/// Provider A: Official/Private API
/// Uses token-based authentication with configurable base URL
class OfficialApiProvider implements PrayerDataProvider {
  late Dio _dio;
  String? _baseUrl;
  String? _apiToken;
  
  @override
  String get id => 'official_api';
  
  @override
  String get name => 'Official Mawaqit API';
  
  @override
  String get description => 'Direct access to official Mawaqit API (requires credentials)';
  
  @override
  List<ConfigField> get configSchema => [
    const ConfigField(
      key: 'baseUrl',
      label: 'API Base URL',
      type: ConfigFieldType.url,
      required: true,
      description: 'e.g., https://mawaqit.net/api',
    ),
    const ConfigField(
      key: 'apiToken',
      label: 'API Token',
      type: ConfigFieldType.password,
      required: true,
      description: 'Your Mawaqit API token',
    ),
    const ConfigField(
      key: 'timeout',
      label: 'Request Timeout (seconds)',
      type: ConfigFieldType.number,
      defaultValue: '30',
    ),
  ];

  @override
  Future<void> initialize(Map<String, dynamic> config) async {
    _baseUrl = config['baseUrl'] as String?;
    _apiToken = config['apiToken'] as String?;
    final timeout = int.tryParse(config['timeout']?.toString() ?? '30') ?? 30;
    
    if (_baseUrl == null || _apiToken == null) {
      throw const ProviderException(
        message: 'Base URL and API token are required',
        providerId: 'official_api',
      );
    }

    _dio = Dio(BaseOptions(
      baseUrl: _baseUrl!,
      connectTimeout: Duration(seconds: timeout),
      receiveTimeout: Duration(seconds: timeout),
      headers: {
        'Authorization': 'Bearer $_apiToken',
        'Accept': 'application/json',
        'User-Agent': 'IqamahApp/1.0',
      },
    ));

    // Add interceptor for logging
    _dio.interceptors.add(LogInterceptor(
      request: true,
      responseBody: true,
      error: true,
    ));
  }

  @override
  Future<List<Mosque>> searchMosques(String query, {GeoLocation? location}) async {
    try {
      final response = await _dio.get('/mosques/search', queryParameters: {
        'q': query,
        if (location != null) 'lat': location.latitude,
        if (location != null) 'lng': location.longitude,
      });

      if (response.statusCode == 200) {
        final List<dynamic> data = response.data['mosques'] ?? [];
        return data.map((json) => _mapMosque(json)).toList();
      }
      
      throw ProviderException(
        message: 'Failed to search mosques: ${response.statusCode}',
        statusCode: response.statusCode,
        providerId: id,
      );
    } on DioException catch (e) {
      throw ProviderException(
        message: 'Network error: ${e.message}',
        statusCode: e.response?.statusCode,
        providerId: id,
      );
    }
  }

  @override
  Future<List<Mosque>> getNearbyMosques(GeoLocation location, {double radiusKm = 10}) async {
    try {
      final response = await _dio.get('/mosques/nearby', queryParameters: {
        'lat': location.latitude,
        'lng': location.longitude,
        'radius': radiusKm,
      });

      if (response.statusCode == 200) {
        final List<dynamic> data = response.data['mosques'] ?? [];
        return data.map((json) => _mapMosque(json)).toList();
      }
      
      throw ProviderException(
        message: 'Failed to get nearby mosques: ${response.statusCode}',
        statusCode: response.statusCode,
        providerId: id,
      );
    } on DioException catch (e) {
      throw ProviderException(
        message: 'Network error: ${e.message}',
        statusCode: e.response?.statusCode,
        providerId: id,
      );
    }
  }

  @override
  Future<PrayerTimes> getPrayerTimes(String mosqueId, {DateTime? date}) async {
    try {
      final targetDate = date ?? DateTime.now();
      final response = await _dio.get('/mosques/$mosqueId/times', queryParameters: {
        'date': '${targetDate.year}-${targetDate.month.toString().padLeft(2, '0')}-${targetDate.day.toString().padLeft(2, '0')}',
      });

      if (response.statusCode == 200) {
        return _mapPrayerTimes(response.data, mosqueId);
      }
      
      throw ProviderException(
        message: 'Failed to get prayer times: ${response.statusCode}',
        statusCode: response.statusCode,
        providerId: id,
      );
    } on DioException catch (e) {
      throw ProviderException(
        message: 'Network error: ${e.message}',
        statusCode: e.response?.statusCode,
        providerId: id,
      );
    }
  }

  @override
  Future<Mosque> getMosqueDetails(String mosqueId) async {
    try {
      final response = await _dio.get('/mosques/$mosqueId');

      if (response.statusCode == 200) {
        return _mapMosque(response.data);
      }
      
      throw ProviderException(
        message: 'Failed to get mosque details: ${response.statusCode}',
        statusCode: response.statusCode,
        providerId: id,
      );
    } on DioException catch (e) {
      throw ProviderException(
        message: 'Network error: ${e.message}',
        statusCode: e.response?.statusCode,
        providerId: id,
      );
    }
  }

  @override
  Future<bool> testConnection() async {
    try {
      final response = await _dio.get('/status');
      return response.statusCode == 200;
    } catch (e) {
      return false;
    }
  }

  Mosque _mapMosque(Map<String, dynamic> json) {
    return Mosque(
      id: json['id']?.toString() ?? '',
      name: json['name'] ?? json['mosque_name'] ?? 'Unknown Mosque',
      address: json['address'] ?? json['localisation'] ?? json['street'],
      city: json['city'] ?? json['ville'],
      country: json['country'] ?? json['pays'],
      latitude: json['latitude'] != null ? double.tryParse(json['latitude'].toString()) : null,
      longitude: json['longitude'] != null ? double.tryParse(json['longitude'].toString()) : null,
    );
  }

  PrayerTimes _mapPrayerTimes(Map<String, dynamic> json, String mosqueId) {
    final date = DateTime.parse(json['date']?.toString() ?? DateTime.now().toIso8601String());
    final times = json['times'] ?? json;

    return PrayerTimes(
      date: date,
      mosqueId: mosqueId,
      cachedAt: DateTime.now(),
      fajr: _parsePrayer('Fajr', times['fajr'] ?? times['Fajr'] ?? times['sobh'], date),
      dhuhr: _parsePrayer('Dhuhr', times['dhuhr'] ?? times['Dhuhr'] ?? times['duhr'], date),
      asr: _parsePrayer('Asr', times['asr'] ?? times['Asr'], date),
      maghrib: _parsePrayer('Maghrib', times['maghrib'] ?? times['Maghrib'], date),
      isha: _parsePrayer('Isha', times['isha'] ?? times['Isha'], date),
      jumuah: times['jumuah'] != null || times['Jumuah'] != null
        ? _parsePrayer('Jumuah', times['jumuah'] ?? times['Jumuah'], date)
        : null,
    );
  }

  Prayer _parsePrayer(String name, dynamic data, DateTime baseDate) {
    if (data == null) {
      return Prayer(name: name, adhan: baseDate, iqama: null);
    }

    if (data is Map) {
      return Prayer(
        name: name,
        adhan: _parseTime(data['adhan'] ?? data['time'] ?? data, baseDate),
        iqama: data['iqama'] != null ? _parseTime(data['iqama'], baseDate) : null,
      );
    }

    // Single time string - assume it's adhan
    return Prayer(
      name: name,
      adhan: _parseTime(data, baseDate),
      iqama: null,
    );
  }

  DateTime _parseTime(dynamic timeData, DateTime baseDate) {
    if (timeData is DateTime) return timeData;
    
    final timeStr = timeData.toString();
    final parts = timeStr.split(':');
    
    if (parts.length >= 2) {
      final hour = int.tryParse(parts[0]) ?? 0;
      final minute = int.tryParse(parts[1]) ?? 0;
      return DateTime(baseDate.year, baseDate.month, baseDate.day, hour, minute);
    }
    
    return baseDate;
  }

  @override
  void dispose() {
    _dio.close();
  }
}

File: src-tauri/src/providers/official_api_provider.rs
use async_trait::async_trait;
use chrono::{NaiveDate, Utc};
use reqwest::Client;
use serde_json::Value;

use crate::models::*;
use crate::providers::{PrayerDataProvider, ProviderError, ProviderResult};

/// Official API Provider (Provider A)
/// Direct Mawaqit API access (requires token)
pub struct OfficialApiProvider {
    client: Client,
    base_url: String,
    api_token: Option<String>,
}

impl OfficialApiProvider {
    pub fn new() -> Self {
        Self {
            client: Client::new(),
            base_url: "https://mawaqit.net/api".to_string(),
            api_token: None,
        }
    }

    fn ensure_initialized(&self) -> ProviderResult<&str> {
        self.api_token.as_deref().ok_or_else(|| {
            ProviderError::InvalidConfig("API token required".to_string())
        })
    }
}

#[async_trait]
impl PrayerDataProvider for OfficialApiProvider {
    fn id(&self) -> &str {
        PROVIDER_OFFICIAL_API
    }

    fn name(&self) -> &str {
        "Mawaqit Official API"
    }

    fn description(&self) -> &str {
        "Direct access to Mawaqit prayer times API (requires API token)"
    }

    fn config_schema(&self) -> Vec<ConfigField> {
        vec![
            ConfigField::new("api_token", "API Token", ConfigFieldType::Password)
                .required()
                .description("Your Mawaqit API token"),
            ConfigField::new("base_url", "Base URL", ConfigFieldType::Url)
                .default_value("https://mawaqit.net/api")
                .description("Optional: Custom API base URL"),
        ]
    }

    async fn initialize(&mut self, config: Value) -> ProviderResult<()> {
        if let Some(token) = config["api_token"].as_str() {
            self.api_token = Some(token.to_string());
        }
        if let Some(url) = config["base_url"].as_str() {
            self.base_url = url.trim_end_matches('/').to_string();
        }
        Ok(())
    }

    async fn search_mosques(&self, query: &str, _location: Option<&GeoLocation>) -> ProviderResult<Vec<Mosque>> {
        let token = self.ensure_initialized()?;

        let response = self.client
            .get(format!("{}/mosques/search", self.base_url))
            .query(&[("word", query)])
            .header("Authorization", format!("Bearer {}", token))
            .send()
            .await
            .map_err(|e| ProviderError::Network(format!("Failed to search: {}", e)))?;

        if !response.status().is_success() {
            return Err(ProviderError::Server {
                status_code: response.status().as_u16(),
                message: "Search failed".to_string(),
            });
        }

        let data: Vec<serde_json::Value> = response.json().await.map_err(|e| {
            ProviderError::Parse(format!("Failed to parse: {}", e))
        })?;

        let mosques: Vec<Mosque> = data
            .into_iter()
            .filter_map(|v| {
                Some(Mosque {
                    id: v.get("id")?.as_str()?.to_string(),
                    name: v.get("name")?.as_str()?.to_string(),
                    address: v.get("address")?.as_str().map(|s| s.to_string()),
                    city: v.get("city")?.as_str().map(|s| s.to_string()),
                    country: v.get("country")?.as_str().map(|s| s.to_string()),
                    latitude: v.get("latitude")?.as_f64(),
                    longitude: v.get("longitude")?.as_f64(),
                    is_favorite: false,
                    last_accessed: None,
                })
            })
            .collect();

        Ok(mosques)
    }

    async fn get_nearby_mosques(&self, location: &GeoLocation, radius_km: f64) -> ProviderResult<Vec<Mosque>> {
        let token = self.ensure_initialized()?;

        let response = self.client
            .get(format!("{}/mosques/nearby", self.base_url))
            .query(&[
                ("lat", location.latitude.to_string()),
                ("lon", location.longitude.to_string()),
                ("radius", radius_km.to_string()),
            ])
            .header("Authorization", format!("Bearer {}", token))
            .send()
            .await
            .map_err(|e| ProviderError::Network(format!("Failed to get nearby: {}", e)))?;

        if !response.status().is_success() {
            return Err(ProviderError::Server {
                status_code: response.status().as_u16(),
                message: "Nearby search failed".to_string(),
            });
        }

        let data: Vec<serde_json::Value> = response.json().await.map_err(|e| {
            ProviderError::Parse(format!("Failed to parse: {}", e))
        })?;

        let mosques: Vec<Mosque> = data
            .into_iter()
            .filter_map(|v| {
                Some(Mosque {
                    id: v.get("id")?.as_str()?.to_string(),
                    name: v.get("name")?.as_str()?.to_string(),
                    address: v.get("address")?.as_str().map(|s| s.to_string()),
                    city: v.get("city")?.as_str().map(|s| s.to_string()),
                    country: v.get("country")?.as_str().map(|s| s.to_string()),
                    latitude: v.get("latitude")?.as_f64(),
                    longitude: v.get("longitude")?.as_f64(),
                    is_favorite: false,
                    last_accessed: None,
                })
            })
            .collect();

        Ok(mosques)
    }

    async fn get_prayer_times(&self, mosque_id: &str, date: Option<NaiveDate>) -> ProviderResult<PrayerTimes> {
        let token = self.ensure_initialized()?;

        let mut request = self.client
            .get(format!("{}/mosques/{}/times", self.base_url, mosque_id));

        if let Some(d) = date {
            request = request.query(&[("date", d.format("%Y-%m-%d").to_string())]);
        }

        let response = request
            .header("Authorization", format!("Bearer {}", token))
            .send()
            .await
            .map_err(|e| ProviderError::Network(format!("Failed to get times: {}", e)))?;

        if response.status().as_u16() == 404 {
            return Err(ProviderError::NotFound(format!("Mosque {} not found", mosque_id)));
        }

        if !response.status().is_success() {
            return Err(ProviderError::Server {
                status_code: response.status().as_u16(),
                message: "Failed to get prayer times".to_string(),
            });
        }

        let data: serde_json::Value = response.json().await.map_err(|e| {
            ProviderError::Parse(format!("Failed to parse: {}", e))
        })?;

        // Parse prayer times from API response
        // This is a simplified implementation - adjust according to actual Mawaqit API
        let date = date.unwrap_or_else(|| Utc::now().date_naive());
        let base_date = date.and_hms_opt(0, 0, 0).unwrap().and_local_timezone(Utc).unwrap();

        fn parse_time(date: chrono::DateTime<Utc>, time_str: &str) -> chrono::DateTime<Utc> {
            let parts: Vec<&str> = time_str.split(':').collect();
            let hour: u32 = parts.get(0).unwrap_or(&"0").parse().unwrap_or(0);
            let minute: u32 = parts.get(1).unwrap_or(&"0").parse().unwrap_or(0);
            date.date_naive().and_hms_opt(hour, minute, 0).unwrap().and_local_timezone(Utc).unwrap()
        }

        let get_time = |key: &str| -> chrono::DateTime<Utc> {
            let time_str = data.get(key).and_then(|v| v.as_str()).unwrap_or("00:00");
            parse_time(base_date, time_str)
        };

        let get_iqama = |key: &str| -> Option<chrono::DateTime<Utc>> {
            data.get(key).and_then(|v| v.as_str()).map(|t| parse_time(base_date, t))
        };

        Ok(PrayerTimes {
            date: base_date,
            fajr: Prayer {
                name: "Fajr".to_string(),
                adhan: get_time("fajr"),
                iqama: get_iqama("fajr_iqama"),
                custom_rakah_count: None,
            },
            dhuhr: Prayer {
                name: "Dhuhr".to_string(),
                adhan: get_time("dhuhr"),
                iqama: get_iqama("dhuhr_iqama"),
                custom_rakah_count: None,
            },
            asr: Prayer {
                name: "Asr".to_string(),
                adhan: get_time("asr"),
                iqama: get_iqama("asr_iqama"),
                custom_rakah_count: None,
            },
            maghrib: Prayer {
                name: "Maghrib".to_string(),
                adhan: get_time("maghrib"),
                iqama: get_iqama("maghrib_iqama"),
                custom_rakah_count: None,
            },
            isha: Prayer {
                name: "Isha".to_string(),
                adhan: get_time("isha"),
                iqama: get_iqama("isha_iqama"),
                custom_rakah_count: None,
            },
            jumuah: None,
            mosque_id: Some(mosque_id.to_string()),
            mosque_name: None,
            cached_at: Some(Utc::now()),
        })
    }

    async fn test_connection(&self) -> ProviderResult<ProviderTestResult> {
        let start = std::time::Instant::now();

        match self.ensure_initialized() {
            Ok(token) => {
                let response = self.client
                    .get(format!("{}/user/me", self.base_url))
                    .header("Authorization", format!("Bearer {}", token))
                    .send()
                    .await;

                match response {
                    Ok(resp) => {
                        let latency = start.elapsed().as_millis() as u64;
                        if resp.status().is_success() {
                            Ok(ProviderTestResult {
                                success: true,
                                message: "Connection successful".to_string(),
                                latency_ms: Some(latency),
                            })
                        } else {
                            Ok(ProviderTestResult {
                                success: false,
                                message: format!("Invalid token: {}", resp.status()),
                                latency_ms: Some(latency),
                            })
                        }
                    }
                    Err(e) => Ok(ProviderTestResult {
                        success: false,
                        message: format!("Connection failed: {}", e),
                        latency_ms: None,
                    }),
                }
            }
            Err(e) => Ok(ProviderTestResult {
                success: false,
                message: format!("Not configured: {}", e),
                latency_ms: None,
            }),
        }
    }

    async fn get_mosque_details(&self, mosque_id: &str) -> ProviderResult<Mosque> {
        let token = self.ensure_initialized()?;

        let response = self.client
            .get(format!("{}/mosques/{}", self.base_url, mosque_id))
            .header("Authorization", format!("Bearer {}", token))
            .send()
            .await
            .map_err(|e| ProviderError::Network(format!("Failed to get details: {}", e)))?;

        if response.status().as_u16() == 404 {
            return Err(ProviderError::NotFound(format!("Mosque {} not found", mosque_id)));
        }

        if !response.status().is_success() {
            return Err(ProviderError::Server {
                status_code: response.status().as_u16(),
                message: "Failed to get mosque details".to_string(),
            });
        }

        let v: serde_json::Value = response.json().await.map_err(|e| {
            ProviderError::Parse(format!("Failed to parse: {}", e))
        })?;

        Ok(Mosque {
            id: v.get("id").and_then(|v| v.as_str()).unwrap_or(mosque_id).to_string(),
            name: v.get("name").and_then(|v| v.as_str()).unwrap_or("Unknown").to_string(),
            address: v.get("address").and_then(|v| v.as_str()).map(|s| s.to_string()),
            city: v.get("city").and_then(|v| v.as_str()).map(|s| s.to_string()),
            country: v.get("country").and_then(|v| v.as_str()).map(|s| s.to_string()),
            latitude: v.get("latitude").and_then(|v| v.as_f64()),
            longitude: v.get("longitude").and_then(|v| v.as_f64()),
            is_favorite: false,
            last_accessed: None,
        })
    }
}
